microsoft/botbuilder-js
# End-to-End Streaming Tests in botbuilder-js

Fixes #2226

## Description
PR automates the testing of botframework-streaming's compatibility with browsers. 

- This is done using a React app that uses DLJS and a bot that both point to the latest preview versions of botframework-streaming. They communicate by enabling DL ASE. 
- Tests are automated using Azure Pipelines as well as leveraging Selenium WebDriver to automate browser testing

## Specific Changes
*E2E Streaming Tests Architecture*
![image](https://user-images.githubusercontent.com/35248895/88363581-6209ed80-cd35-11ea-8312-3c7a5f3f5303.png)


[E2E Streaming Tests Pipeline here](https://fuselabs.visualstudio.com/SDK_v4/_build?definitionId=1018&_a=summary) (currently under temporary name "[experimental]E2E_Streaming_Test_Ashley")

Running the pipeline will:
- **Provision & Deploy Projects**:
    - **Bot**: point bot to use the daily build of botframework-streaming (latest preview version)
    - **React App**:
        - Provision a brand new react app via `npx create-react-app` command
        - Set up the React app to use WebChat (`ReactWebChat`)
        - Install a custom version of DirectLineJS (DLJS) into the app
            - Configures DLJS to use the latest botframework-streaming preview version
- **Run Tests**
    - ~Downloads the browser-appropriate Driver necessary for Selenium WebDriver to automate testing in a browser (Chrome)~ Use the windows-2019 image's pre-installed of WebDriver & ChromeDriver to automate testing in browser
    - Runs the tests that have React and bot talk to each other

See [README](https://github.com/microsoft/botbuilder-js/blob/Zerryth/streaming-e2e-tests/libraries/streaming-e2e-tests/README.md) for further details of PR, if wanted.

## Testing


It's all green!
![image](https://user-images.githubusercontent.com/35248895/88363782-02f8a880-cd36-11ea-943c-3ef861c3bc85.png)

___

This PR is already large enough, and it only tests 1 scenario (webpack'd project + chrome browser). In future expansions of streaming tests, we could include the following (detailed per William):

*Different Build Scenarios*:
1. Use `create-react-app` (CRA) without WebChat, just talk to bot using DLJS
2. Use webpack without CRA
3. Add Rollup in addition to webpack
 
*Different Browsers*:
- Edge UWP
- Edge Chromium
- Firefox
- IE11
- macOS Safari
- iOS Safari
- Android Chrome

The expansion should go more quickly now that this PR lays down the foundational infrastructure for E2E testing of streaming bits (at least for adding additional browsers to test against).
